### PR TITLE
Replica Validation Refactor

### LINKED ISSUE #2090: Review and refactor state validation logic
Some quick fixes to `validate_and_apply_header` in #2089 have left it in a disorganized state. We now have some `async` calls which need to be dealt with, but the primary validation function is synchronous. We should think about the best way to organize this. Ideally each validation step would be represented by a variant of `ProposalValidationError` and in `validate_and_apply_header` we would log those as we currently do with `validate_proposal`.

### PR DESCRIPTION
**Pull request by tbro**
Closes #2090 

### This PR:
Refactors contents of `validate_and_apply_header` by introducing two new internal (non-pub) types. This should facilitate:

  * comprehension
  * documentation
  * testing

Basically this PR:

  * adds `Proposal` and `ValidatedTransition` types
  * moves validation steps to isolated methods.
  * adds doc comments to methods
  * updates error enum
  * adds many tests

## Review

Most important changes are isolated `./types/src/v0/impls/state.rs`. Many unit tests have been added there. Tests should pass and you should be convinced that there is sufficient coverage. Small changes were made to `./types/src/v0/impls/header.rs` where we previously had testing for this validation in the form of high level *success cases* and *eror cases* tests. The success case test has been renamed and the error case test removed in favor of the more comprehensive error case tests in the new unit tests. 

## Review Documentation

You can run `just doc` and navigate here in your browser:
`./target/doc/espresso_types/v0/impls/state/struct.ValidatedTransition.html`

### CODE REVIEW

**Review by zacshowa — COMMENTED** [review]
I know this is a WIP, but I left some general discussion points from my first read. I think that this will be a great use of the type system to make validation more readable.

File: types/src/v0/impls/state.rs
Comment (tbro) [reply types/src/v0/impls/state.rs:386]: Generally speaking, I think you want the things you can invalidate the most quickly to come first. This way you avoid doing a bunch of work on steps that pass, only to invalidate at the end. There are other considerations though. Timestamps are time sensitive, so validation could potentially be affected by steps that take time. Fee validation requires network calls and waiting. So some waiting is required for some validation steps. There is similar complexity to catchup. So there is some complexity we have to navigate.

File: types/src/v0/impls/state.rs
Comment (tbro) [reply types/src/v0/impls/state.rs:482]: Requires some though. Putting it in chain config means we require an upgrade to modify it. Maybe we can just pass it in on the command line.

File: types/src/v0/impls/state.rs
Comment (tbro) [reply types/src/v0/impls/state.rs:865]: I don't believe `l1_client` makes any network connections on its own... you are required to call one of the provided methods for it to connect to anything. I also don't know if we ultimately want this clone. We could possibly pass `&NodeState` at appropriate places. Also keep in mind the types here are pretty hap hazard so far. We may still change some things, so probably not worth going down a rabbit hole over this. But before it gets merged we definitely need to pin down all the shapes.

File: types/src/v0/impls/state.rs
Comment (zacshowa) [reply types/src/v0/impls/state.rs]: Ah, I wasn't aware the chain config was immutable between upgrades. What's the reasoning for that?

File: types/src/v0/impls/state.rs
Comment (zacshowa) [reply types/src/v0/impls/state.rs:913]: I agree that it's fine for now and as things settle a bit more we can work on optimizing that front. It seems like ValidatedTransition doesn't need many mutable borrows, so we likely don't need to take ownership over most of the data it holds. Then the question becomes memory trade off vs the complexity of reading lifetimes.

File: types/src/v0/impls/state.rs
Comment (zacshowa) [reply types/src/v0/impls/state.rs:438]: Hmm. This makes sense. Thanks for the insight!

File: types/src/v0/impls/state.rs
Comment (tbro) [reply types/src/v0/impls/state.rs]: It's meant to be immutable configuration for the block_chain. Part of blockchain the safety concerns I think.

File: types/src/v0/impls/state.rs
Comment (tbro) [reply types/src/v0/impls/state.rs]: I think so. 1ccf3c092a7e42cdfc04fb671b5a281a383758a1

File: types/src/v0/impls/header.rs
Comment (sveitser) [inline types/src/v0/impls/header.rs:1494]: Is this intentionally left commented out?

File: types/src/v0/impls/state.rs
Comment (sveitser) [inline types/src/v0/impls/state.rs:325]:
```suggestion
    /// Validates [`ChainConfig`] of proposal is equal to the one stored in
    /// state by comparing commitments.
```

File: types/src/v0/impls/state.rs
Comment (sveitser) [inline types/src/v0/impls/state.rs:390]: A bit of a nitpick: the transition is only validated if `validate()` does not return an error but given current naming one might expect it to be validated upon creation unless one actually inspects the return type. I'm totally fine with the current name just thinking maybe we can make it even clearer at some point.

File: types/src/v0/impls/state.rs
Comment (sveitser) [inline types/src/v0/impls/state.rs:560]: Nit: comment is now potentially misleading because in this function the drift check is done last.

File: types/src/v0/impls/state.rs
Comment (sveitser) [inline types/src/v0/impls/state.rs:310]: There is also a Proposal type in hotshot that we use in this repo, so I think we can use a different name here. `ProposalValidationInput`? A bit of a mouthful :roll_eyes:

File: types/src/v0/impls/state.rs
Comment (sveitser) [inline types/src/v0/impls/state.rs:483]:
```suggestion
        self.validate_l1_finalized()?;
```

File: types/src/v0/impls/state.rs
Comment (sveitser) [inline types/src/v0/impls/state.rs:776]: Should we do the fee charging as part of `apply_proposal`?

File: types/src/v0/impls/state.rs
Comment (sveitser) [inline types/src/v0/impls/state.rs:1097]:
```suggestion
        /// Replaces builder signature w/ a valid one from a random key pair
```
?

File: types/src/v0/impls/state.rs
Comment (sveitser) [inline types/src/v0/impls/state.rs:1152]: Remove async?

File: types/src/v0/impls/state.rs
Comment (sveitser) [inline types/src/v0/impls/state.rs:1859]: Commented block

**Review by sveitser — APPROVED** [review]
LGTM. I think this is a great improvement in terms of readability. Just some suggestions.

I also think @0xkato should review this PR as well.

File: types/src/v0/impls/header.rs
Comment (tbro) [reply types/src/v0/impls/header.rs:1494]: Humm I thought this was previously commented in relation to the above TODO... Maybe I got confused will double check.

File: types/src/v0/impls/state.rs
Comment (tbro) [reply types/src/v0/impls/state.rs:395]: Its a tough one. It's a type that only exists to facilitate a state transition by encapsulating validation logic. The state is actually held in `ValidatedState`. Open to ideas, but I don't have any at the time being.

File: types/src/v0/impls/state.rs
Comment (tbro) [reply types/src/v0/impls/state.rs:310]: :thinking:

File: types/src/v0/impls/state.rs
Comment (tbro) [reply types/src/v0/impls/state.rs]: 092eba7ec44db27fd2d1fda28d8b7f53df5f800c

File: types/src/v0/impls/state.rs
Comment (tbro) [reply types/src/v0/impls/state.rs]: bf5c57228d3936dda0de969f2cc80e158abb313a

File: types/src/v0/impls/state.rs
Comment (tbro) [reply types/src/v0/impls/state.rs:1109]: The intent is convey that signature validation will fail because we are adding a signature that doesn't match the fee_info. It is therefore invalid (relative to what it will be validated against).

File: types/src/v0/impls/state.rs
Comment (tbro) [reply types/src/v0/impls/state.rs]: cea4a80a5beb971559ee661cc4af193233eb1e9f

File: types/src/v0/impls/state.rs
Comment (tbro) [reply types/src/v0/impls/state.rs]: 104b12150f288edcd801e9e2fc3a216808b894e1

File: types/src/v0/impls/state.rs
Comment (tbro) [reply types/src/v0/impls/state.rs:1104]: Thinking again I see your point... The signature is structurally valid but will fail verification.

**Pull request merged**
